grease-cli/src/Grease/Main.hs

```diff
@@ -598,8 +630,23 @@ simulateMacawCfg la bak fm halloc macawCfgConfig archCtx simOpts setupHook mbCfg
   let mdEntryAbsAddr = fmap (segoffToAbsoluteAddr memory) mbCfgAddr
   initArgs_ <- minimalArgShapes bak archCtx mdEntryAbsAddr
   let argNames = fmapFC (Const . getValueName) rNameAssign
+  -- Is there a better way to do this resolution? We need the pc absent any position independent loading
+  -- so we find the segment that corresponds by checking for all segments if segindex->memsegment -> the segment of the target addr
+  let segIndexToSegment = MM.memSegmentIndexMap memory
+  let targetSegment = fst <$> ((\seg -> find (\(_, seg2) -> seg == seg2) (Map.toList segIndexToSegment)) =<< (MM.segoffSegment <$> mbCfgAddr))
+  let mbSegment =
+        ( do
+            bin <- mcElf macawCfgConfig
+            find (\phdr -> (Just $ Elf.phdrSegmentIndex phdr) == targetSegment) $ headerPhdrs bin
+        )
+  let mbNewAddr =
+        ( \(segHdr, entAddr) ->
+            (MM.memWordValue $ MM.segoffOffset entAddr) + fromIntegral (Elf.phdrSegmentVirtAddr segHdr)
+        )
+          <$> ((,) <$> mbSegment <*> mbCfgAddr)
```

Comment:
Would it perhaps be worth wrapping this logic up as a function that lives beside loadDwarfPreconditions? It could even return a newtype to make it clear how the two are connected.